torch/distributed/device_mesh.py

```diff
@@ -220,7 +226,7 @@ def __init__(
 
             # private field to pre-generate DeviceMesh's hash
             self._flatten_mesh_list = tuple(self.mesh.flatten().tolist())
-            self._hash = hash((self._flatten_mesh_list, self.mesh.shape, id(self)))
+            self._hash = hash((self._flatten_mesh_list, self.mesh.shape))
```

Comment:
Hmm should this also consider `device_type` and `mesh_dim_names`?

Reply:
Great catch. I think 'device_type' should be considered as well, but 'mesh_dim_names' should be captured by taking into consideration of parent_mesh?

Reply:
Ah I see. For the same 2D mesh with different mesh dim names, for example, 

```
mesh_2d = init_device_mesh('cuda', (2, 4), mesh_dim_names=('dp', 'tp')
another_mesh_2d = init_device_mesh('cuda', (2, 4), mesh_dim_names=('pp', 'dp')
```
these two meshes should be considered as different meshes. Is this the reason that you think `mesh_dim_names` should be taken into consideration while generating hash?

Reply:
Yes

Reply:
Hmm. I am thinking whether this would be too restrictive. 

As we discussed, the first two should be considered different as they have different mesh_dim_names. However, if I create a new DeviceMesh without mesh_dim_names, should it be considered the same or different from the first two? I am not sure about this yet. 
```
mesh_2d = init_device_mesh('cuda', (2, 4), mesh_dim_names=('dp', 'tp'))
another_mesh_2d = init_device_mesh('cuda', (2, 4), mesh_dim_names=('pp', 'dp'))
mesh_2d_no_dim_names = init_device_mesh('cuda', (2, 4))
```